### PR TITLE
Add a raycast call to the webgl fat lines code

### PR DESCRIPTION
**Pull request by nathanieltagg**
I find myself in need of a reliable thick-line WebGL implementation, and this one seems the one that I can make work.. it only lacks a raycaster.  I've written my own here, and tested with private code.

### CODE REVIEW

File: examples/js/lines/LineSegments2.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ THREE.LineSegments2 = function ( geometry, material ) {
 
 	this.geometry = geometry !== undefined ? geometry : new THREE.LineSegmentsGeometry();
 	this.material = material !== undefined ? material : new THREE.LineMaterial( { color: Math.random() * 0xffffff } );
-
+  this.raycast_to_segment = true; // False if you don't care which segment is cast.
```
Comment (WestLangley) [inline examples/js/lines/LineSegments2.js]: For consistency, let's not add this.

File: examples/js/lines/LineSegments2.js [hunk 2]
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
 
 		return this;
 
-	}
+	}, 
+  
+  // NJT added:
```
Comment (WestLangley) [inline examples/js/lines/LineSegments2.js]: Thanks for this, but such comments are not required.

File: examples/js/lines/LineSegments2.js [hunk 3]
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
...
+			ray.copy( raycaster.ray ).applyMatrix4( inverseMatrix );
+
+			var localPrecision = precision / ( ( this.scale.x + this.scale.y + this.scale.z ) / 3 );
+			var localPrecisionSq = localPrecision * localPrecision;
+
+			var vStart = new THREE.Vector3();
+			var vEnd = new THREE.Vector3();
+			var interSegment = new THREE.Vector3();
+			var interRay = new THREE.Vector3();
+			var step = ( this && this.isLineSegments ) ? 2 : 1;
```
Comment (WestLangley) [inline examples/js/lines/LineSegments2.js]: `step` is not used.

File: examples/js/lines/LineSegments2.js [hunk 4]
```diff
@@ -60,6 +60,76 @@ THREE.LineSegments2.prototype = Object.assign( Object.create( THREE.Mesh.prototy
...
+
+      // Always a LineSegments2 geometry, which stores positions NOT in the positions attribute NJT
+      var starts = geometry.attributes.instanceStart;
+      var ends   = geometry.attributes.instanceEnd;
+      
+      for(var i=0;i<starts.count;i++) {
```
Comment (WestLangley) [inline examples/js/lines/LineSegments2.js]: Please see https://zz85.github.io/mrdoobapproves/ for help correcting your code style.

**Comment by WestLangley** [comment]
Thank your for your contribution. Please also test on really fat lines. I expect there will be problems, but this approximation may be acceptable.

**Comment by Mugen87** [comment]
@nathanieltagg It would be great if you can implement @WestLangley's feedback since users in the forum requested this feature multiple times:

https://discourse.threejs.org/t/raycasting-with-linesegment2-objects/4256

**Comment by nathanieltagg** [comment]
Hi, I've responded to the code changes as requested.

I've decided to add one more enhancement, a new material called PerspectiveLineMaterial.  This draws the line with a thickness that varies with the distance from the camera. This is a nice way of drawing simple frame structures without needing full 3d tubes. I've been using this in my own code.

I've also updated the fatlines demo to preview the new material, and added a raycast marker to show that it works.

**Comment by nathanieltagg** [comment]
(Sorry if I've done something wrong; this is my first pull request ever.)

**Comment by nathanieltagg** [comment]
OK, I think I've fixed it. I've also added a new feature I think is pretty useful: a perspective line.

— Nathaniel

**Pull request closed without merging**